File: SNACKS/ClientApp/src/app/productos/producto.ts
import { IUnidad } from "../unidades/unidad";
import { ICategoria } from "../categorias/categoria";

export interface IProducto {
  idProducto?: number;
  nombre?: string,
  categoria?: ICategoria;
  esInsumo?: boolean;
  esProducto?: boolean;
  esGasto?: boolean;
  items?: IItemProducto[];
  insumos?: IInsumoProducto[];
}

export interface IItemProducto {
  idItemProducto?: number;
  producto?: IProducto;
  unidad?: IUnidad;
  factor?: number;
}

export interface IInsumoProducto {
  idInsumoProducto?: number;
  idProducto?: number;
  insumo?: IProducto;
}

File: SNACKS/Models/Producto.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace SNACKS.Models
{
    public class Producto
    {
        public int IdProducto { get; set; }
        public string Nombre { get; set; }
        public bool EsProducto { get; set; }
        public bool EsInsumo { get; set; }
        public bool EsGasto { get; set; }
        public Categoria Categoria { get; set; }
        public IList<ItemProducto> Items { get; set; }
        public IList<InsumoProducto> Insumos { get; set; }

        public Producto()
        {
            Items = new List<ItemProducto>();
            Insumos = new List<InsumoProducto>();
        }
    }
}
